ocaml/parsing/parser.mly

```diff
@@ -173,11 +175,22 @@ guaranteed to be used. *)
 let mkpat_with_modes modes pat =
   (* Mark ghost to pass ppxlib well-nestedness check. *)
   let modes = Mode.ghostify modes in
+  
+  (* [pat] might already contain mode annotations (mostly due to mixing legacy
+  and new mode syntax), which is OK. Conceptually, we should create a new
+  [Ppat_mode_constraint] wrapping around the [pat]. But we are not allowed to
+  change parsetree. So the second best is probably to use [Ppat_alias] with mode
+  annotation wrapping around [pat].
+
+  Here, we cheat a bit, and simply merge with the exising mode annotation.
+  *)
```

Comment:
This is complicated. I notice you just added the new syntax into `let_pattern` and `label_let_pattern`. Can we just return a tuple from these productions (where the second component is modes)? That would allow us to combine the mode expressions with the legacy modes by concatentation, rather than the current approach of blindly merging if the structure looks plausibly OK.